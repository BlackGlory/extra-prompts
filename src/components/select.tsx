import { Box, Text } from 'ink'
import { Select as SelectInput } from '@inkjs/ui'
import { IOption } from '@src/types.js'

interface ISelectOptions<T> {
  message: string
  options: IOption<T>[]

  onSelect: (option: IOption<T>) => void
}

export function Select<T>(props: ISelectOptions<T>) {
  const { message, options, onSelect } = props

  return (
    <Box>
      <Text>{message}: </Text>
      <SelectInput
        options={options.map((option, i) => ({
          label: option.label
        , value: i.toString()
        }))}
        onChange={indexString => {
          const index = Number(indexString)
          const option = options[index]

          onSelect(option)
        }}
      />
    </Box>
  )
}
