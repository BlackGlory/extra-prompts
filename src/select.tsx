import { render } from 'ink'
import { Select, IOption } from '@components/select'

export { IOption } from '@components/select'

export function select<T>(message: string, options: IOption<T>[]): Promise<T> {
  return new Promise(resolve => {
    const { unmount } = render(
      <Select
        message={message}
        options={options}
        onSelect={option => {
          unmount()
          resolve(option.value)
        }}
      />
    )
  })
}
