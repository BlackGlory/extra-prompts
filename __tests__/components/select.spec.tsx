import { Select } from '@components/select'
import { render } from 'ink-testing-library'
import { delay } from 'extra-promise'

describe('Select', () => {
  test('render', () => {
    const fn = jest.fn()
    const { lastFrame } = render(
      <Select
        message='message'
        onSelect={fn}
        options={[
          { label: 'Option 1', value: 1 }
        , { label: 'Option 2', value: 2 }
        ]}
      />
    )

    expect(lastFrame()).toMatchSnapshot()
    expect(fn).toBeCalledTimes(0)
  })

  test('onSelect', async () => {
    const fn = jest.fn()
    const { lastFrame, stdin } = render(
      <Select
        message='message'
        onSelect={fn}
        options={[
          { label: 'Option 1', value: 1 }
        , { label: 'Option 2', value: 2 }
        ]}
      />
    )

    await delay(100)
    stdin.write('\u001b[B') // down
    await delay(100)
    stdin.write('\r')
    await delay(100)

    expect(lastFrame()).toMatchSnapshot()
    expect(fn).toBeCalledTimes(1)
    expect(fn).toBeCalledWith({
      label: 'Option 2'
    , value: 2
    })
  })
})
