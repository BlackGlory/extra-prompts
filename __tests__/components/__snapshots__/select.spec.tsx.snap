// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select onSelect 1`] = `
"message:   Option 1
         [34m‚ùØ[39m [34mOption 2[39m"
`;

exports[`Select render 1`] = `
"message: [34m‚ùØ[39m [34mOption 1[39m
           Option 2"
`;
